app/scripts/metamask-controller.js

```diff
@@ -2870,6 +2899,34 @@ export default class MetamaskController extends EventEmitter {
               }
             });
 
+            const addedSolanaScope =
+              sessionScopes[MultichainNetworks.SOLANA] ||
+              sessionScopes[MultichainNetworks.SOLANA_DEVNET] ||
+              sessionScopes[MultichainNetworks.SOLANA_TESTNET];
+
+            if (addedSolanaScope) {
+              const solanaAccountChangedNotifications =
+                currentValue.get(origin)?.sessionProperties?.[
+                  KnownSessionProperties.SolanaAccountChangedNotifications
+                ];
+
+              if (solanaAccountChangedNotifications) {
+                const { accounts } = addedSolanaScope;
+                const solanaFormattedAccounts = accounts.map(
+                  (caipAccountId) => {
+                    const { address } = parseCaipAccountId(caipAccountId);
+                    return address;
+                  },
+                );
+                const accountAddressToEmit = solanaFormattedAccounts.includes(
+                  lastSelectedSolanaAccountAddress,
+                )
+                  ? lastSelectedSolanaAccountAddress
+                  : solanaFormattedAccounts[0];
```

Comment:
This logic allows for the following scenario.
Assume solana accounts A, B, C
A dapp is permitted to use Account A and B
The currently selected account in the wallet is B
The dapp last got an accountChanged event for Account B
The user changes the currently selected account in the wallet to Account C
The dapp will get an accountChanged event for Account A even though it shouldn't have gotten any event at all or at minimum the value should have stayed Account B